src/analyses/loopfreeCallstring.ml

```diff
@@ -0,0 +1,88 @@
+(** This approach is inspired by 
+    @see <https://arxiv.org/abs/2301.06439> Schwarz, M., Saan, S., Seidl, H., Erhard, J., Vojdani, V. Clustered Relational Thread-Modular Abstract Interpretation with Local Traces. Appendix F. *)
+
+open Analyses
+
+module Spec : MCPSpec =
+struct
+  include Analyses.IdentitySpec
+
+  let name () = "loopfree_callstring"
+
+  module FundecSet = SetDomain.Make (CilType.Fundec)
+  module FundecList = Printable.Liszt (CilType.Fundec) 
+  module Either = struct 
+    include Lattice.Flat (Printable.Either (FundecList) (FundecSet)) (* should be a list. Since a Lattice is required, Lattice.Flat is used to fulfill the type *) 
+
+    let printXml f = function
+      | `Lifted x -> 
+        begin
+          match x with 
+          | `Left x -> BatPrintf.fprintf f "<value>stack:\n%a</value>" FundecList.printXml x
+          | `Right x -> BatPrintf.fprintf f "<value>set:\n%a</value>" FundecSet.printXml x
+        end;
+      | _ -> failwith "Error loopfreeCallstring (printXml): the Flat Lattice containing Either shouldn't be `Top or `Bottom!"
+  end
+  module D = Lattice.Liszt (Either)
```

Comment:
This reimplements the prefix-set data structure we already have in `ThreadIdDomain.History`. We should extract a common functor which can be reused for both to minimize the duplication.

Actually, the structure here is different: it's a list of eithers which could contain multiple lists and multiple sets. Is that intentional? If so, then what's the meaning of that? It wasn't so in the ESOP paper. If not, then this should just use the `ThreadIdDomain` structure whose type enforces that there aren't multiple lists or sets mixed.

Is there a reason this cannot use anything other than `fundec`s?
Conceptually, it's yet another stack abstraction like k-bounded stacks, etc. I think the prefix-set abstraction should fit into the other callstring analysis being proposed here.
That's what has been already done with the old analyses in `StackDomain`.

Reply:
Yes it is intentional that sets and lists (now I updated it to elements) are mixed in one list. The Loopfree callstring approach is similar to the call string approach and stores the call stack containing the prior function calls. Our idea was to improve this approach by storing any loop occurring in this call stack by a set. So in case of the call-sequence `main -> f -> f` the first call to `f` has the context as for the call string approach `(main, f)`, and the second call to `f` has the context `(main, {f})`, where the set represents the recursive call. In case the call-sequence has more `f`s in it, the context of all those comming `f` is still `(main, {f})` and consequently, the number of contexts is reduced. And this is basically the goal, to reduce the number of contexts during analyzation to improve runtime and hopefully not loose to much precision. 
Also for bigger loops this would hold: `main -> f -> g -> h -> f` would behave similar to the call string approach till the second function call of `f`. For this second call, the context `(main, {f, g, h})` would be the context.

Reply:
The current idea was to make it similar to the call string approach of Sharir and Pnueli, therefore only the `fundecs` are used as elements. Maybe one can think about using the statements as alternative input and use it as a second approach? @jerhard do you think this would make sense?

Reply:
The examples you gave now are all pairs of a prefix and a set, which is why I would expect the type to also be such a pair. My point is that the current type also contains values like `[[main], {f, g, h}, [foo, bar], [baz], {ping, pong}]` because it's a list of either prefix or set. Are such things actually used?

Reply:
> Conceptually, it's yet another stack abstraction like k-bounded stacks, etc. I think the prefix-set abstraction should fit into the other callstring analysis being proposed here.
> That's what has been already done with the old analyses in StackDomain.
I am currently doing some changes on loopfree, and can check if it would be possible to reuse the `StackDomain`

Reply:
Yes they are all used. For example for the call-sequence: `main -> f -> f -> g -> g` we would compute for the last `g` the context `[main, {f}, {g}]`. Or if we don't have any recursion `main -> f -> g` it would result in `[main, f, g]`.

Reply:
In the current code the stack only contains `Fundecs` and sets of `Fundecs`, since this makes the implementation simpler and more efficient. So there are no lists anymore inside the stack

Reply:
Oh, so this is a generalization of the history-based thread IDs we used in the ESOP paper. Even so, both variations should fit the same abstraction of stacks. That would allow using this generalization also for thread IDs.

If we've come up and implemented all these different stack abstractions, it would be a shame to needlessly hard-code each one to a different element type or use case.

Reply:
> the tids do not just abstract the creation stack, but also have this C component of encountered create edges

Those are just part of the `threadid` analysis domain. `ThreadIdDomain.History`, which is the stack abstraction has nothing about edges.

Reply:
> which is the stack abstraction has nothing about edges

It does not directly track this information but the logic is in `ThreadidDomain.History` nevertheless (but not in `compose`). 

~~~OCaml
  let threadenter ~multiple ((p, _ ) as current, (cs,_)) (n: Node.t) i v =
    let ni = Base.threadenter ~multiple n i v in
    let ((p', s') as composed) = compose current ni in
    if is_unique composed && (S.mem ni cs || multiple) then
      [(p, S.singleton ni); composed] (* also respawn unique version of the thread to keep it reachable while thread ID sets refer to it *)
    else
      [composed]
~~~

All I am saying is that we have to see how much re-use makes sense without the code getting overly complicated. The creation reasoning for tids for example is already quite involved, so I am not sure if generalizing that is a good idea.

Reply:
That's true, history thread IDs require a few more operations than `compose`/`add` for the stack abstraction: `is_unique`, `is_must_parent`, `may_create`, but these can probably be recast as generic stack properties.
There's no need to do that here though. I now understand how this abstraction is different from the history thread IDs one.